datadog/dogstatsd/aggregator.py

```diff
@@ -0,0 +1,75 @@
+import threading
+from datadog.dogstatsd.metrics import (
+    CountMetric,
+    GaugeMetric,
+    SetMetric,
+)
+from datadog.dogstatsd.metric_types import MetricType
+
+
+class Aggregator(object):
+    def __init__(self, client):
+        self.client = client
+        self.metrics_map = {
+            MetricType.COUNT: {},
+            MetricType.GAUGE: {},
+            MetricType.SET: {},
+        }
+        self._locks = {
+            MetricType.COUNT: threading.RLock(),
+            MetricType.GAUGE: threading.RLock(),
+            MetricType.SET: threading.RLock(),
+        }
+
+    def start(self, flush_interval):
+        self.flush_interval = flush_interval
+        self.client._start_flush_thread(flush_interval, self.send_metrics)
```

Comment:
This is looking good. One more design suggestion that may further simplify things:

Currently `Aggregator` takes an instance of the client. Presumably, the aggregator instance will be created by the client during startup (correct me if I am wrong) - this will create a dependency cycle. (which is not a "problem" in python, but could make testing because the components are tightly coupled). 

It may make the code simpler if the client made calls to the aggregator and the aggregator was not aware of the client at all. For example:

- you can remove the `start` and `stop` methods here since the flush thread is managed by the client. 
- `send_metrics` could instead return the list of metrics to flush and `client._report` can be executed from within the client. 

This would also simply writing tests since the aggregator simply takes metrics in, and aggregated metrics come out - and the threading/lifecycle is completely controlled by the client. 

Let me know what you think or if you have any questions. I think this should be a relatively small code change considering the state this change is in right now :)

Reply:
Yeah this makes sense, thanks!